### PR TITLE
fix(amf): bug fix for the issue 9257

### LINKED ISSUE #9257: 5g SA | TeraVM |. Downlink traffic breaks after Paging procedure & UE turns to connected state
Scenario: 
1, Registration and Establish PDU session.
2, UL and DL traffic is good with NAT mode enabled in AGW.
3, UE goes to idle Mode with Context release
4, AGW initiate Paging
5, UE sends the Service Request with service type: Mobile terminated Services (02)
6, [Issues] AGW receives downlink traffic packets in N6 to natted IP but the pkts are not encapsulated into GTP tunnel in N3 Interface

Tested Build: 1.7.0-1632278348-9baf09db
Earlier Success: Since the merge this feature fails in user plan after Paging procedure where as Control plane flows looks good.

[Traces]
[https://app.zenhub.com/files/170803235/e7309106-cf43-4c39-91aa-70de8a35aabe/download](https://app.zenhub.com/files/170803235/e7309106-cf43-4c39-91aa-70de8a35aabe/download)

Logs: Need Developers Inputs to collect most relevant logs ? , ex:  OVS Debug, table0 ??

Expectation:
After Paging procedure and UE turns to connected the packets received in N6 should get into GTP tunnel & reach UE.

Snapshot:

### PR DESCRIPTION
**Pull request by VinayKumar1591**
fix(amf): bug fix for the issue 9257

## Summary

1. Changes to send GNB tunnel info to SMF
2. Not sending PduSesssionResourceSetupTransferIE from InitialContextSetupReq if service type is signaling
3. Changes to resolve Pdu Session establishment Request NAS Decoding failure on NGAP Tester

## Test Plan

1. Tested on NGAP Tester

### CODE REVIEW

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp [hunk 1]
```diff
@@ -395,12 +395,13 @@ int amf_idle_mode_procedure(amf_context_t* amf_ctx) {
...
+  smf_context_t* smf_ctx = nullptr;
+  int j                  = 0;
 
   for (auto it = ue_context_p->amf_context.smf_ctxt_vector.begin();
-       it != ue_context_p->amf_context.smf_ctxt_vector.end(); it++) {
-    smf_ctx                   = *it;
-    smf_ctx.pdu_session_state = INACTIVE;
+       it != ue_context_p->amf_context.smf_ctxt_vector.end(); it++, j++) {
+    smf_ctx = ue_context_p->amf_context.smf_ctxt_vector.data() + j;
+    smf_ctx->pdu_session_state = INACTIVE;
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp]: Looks quite convoluted. Why don't you simply use `it->pdu_session_state = INACTIVE`?

**Review by ulaskozat — CHANGES_REQUESTED** [review]
see the comments

File: lte/gateway/c/core/oai/tasks/nas5g/src/M5GPDUSessionEstablishmentRequest.cpp [hunk 2]
```diff
@@ -115,8 +115,12 @@ int PDUSessionEstablishmentRequestMsg::DecodePDUSessionEstablishmentRequestMsg(
           decoded += decoded_result;
         }
         break;
-      case REQUEST_5GSM_CAPABILITY_TYPE:
       case REQUEST_MAXIMUM_NUMBER_OF_SUPPORTED_PACKET_FILTERS_TYPE:
+        type_len = sizeof(uint8_t);
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/nas5g/src/M5GPDUSessionEstablishmentRequest.cpp:118]: add/modify unit tests for encoding/decoding.

Comment (VinayKumar1591) [reply lte/gateway/c/core/oai/tasks/nas5g/src/M5GPDUSessionEstablishmentRequest.cpp:118]: Done

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp [hunk 3]
```diff
@@ -298,6 +304,23 @@ TEST(test_amf_nas5g_pkt_process, test_amf_pdu_sess_est_req_type2_msg) {
...
+  uint16_t buf_len     = 1024;
+
+  uint32_t len = nas5g_pkt_snap.get_pdu_session_est_type3_len();
+
+  /* Check if uplink pdu packet is parsed properly */
+  memset(&pdu_sess_est_req, 0, sizeof(ULNASTransportMsg));
+  decode_res = decode_ul_nas_transport_msg(
+      &pdu_sess_est_req, nas5g_pkt_snap.pdu_session_est_req_type3, len);
+
+  EXPECT_EQ(decode_res, true);
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf.cpp:321]: this does not do much. you should compare against the expected fields.

Comment (ulaskozat) [reply lte/gateway/c/core/oai/test/amf/test_amf.cpp:321]: @VinayKumar1591 can you submit another set of PRs that actually perform both encoding and decoding for amf test cases?

**Review by ulaskozat — APPROVED** [review]
LGTM. Please implement the encoding functions in another PR and update the test cases where we start with the data structure, encode it to a buffer, decode the buffer to another variable of the same data structure, then do memcmp for testing.

**Pull request merged**